### DISCUSSION TITLE
Single executable applications - WebAssembly target

### DISCUSSION THREAD

**Discussion by krpeacock** [nodejs/help#4721] [Ideas]
Hello maintainers,

I hope you are doing well. I would like to propose adding support for WebAssembly / Wasmtime as a build target in Node.js as an extension of the new [single executable applications](https://nodejs.org/api/single-executable-applications.html) feature.

As you may know, WebAssembly is a binary format that enables us to run code written in multiple programming languages, including C++, Rust, and Go, in a cross-platform environment. It provides an efficient, secure, and portable way to run code across different platforms.

By adding support for WebAssembly / Wasmtime as a build target, Node.js devs could leverage its benefits to open up new cross-platform opportunities. Here are a few of the advantages:

1. Cross-platform compatibility: WebAssembly code can be run on various platforms, including Windows, macOS, Linux, and even the browser. This means we can target a wide range of platforms with a single build, which simplifies the build and deployment process.
2. High performance: WebAssembly is designed to be fast and efficient, which makes it ideal for running code that requires high-performance, such as machine learning or scientific computing.
3. Security: WebAssembly code runs in a sandboxed environment, which provides an additional layer of security against malicious code.

I believe adding support for WebAssembly / Wasmtime as a build target would be an interesting step  forward for Node.js and WebAssembly, and I hope you consider it for future releases.

Thank you for your attention, and please let me know if you have any questions or concerns.

Sincerely,
Kyle

**Comment by krpeacock** [comment]
WebAssembly would probably be best suited for a subset of Node.js features - not everything Node.js can do in an OS context is possible in wasm, and having a lightweight module is typically preferable. Looking forward to the discussion!

**Comment by mhdawson** [comment]
@krpeacock I think you could already do this.  You can write a script that loads/runs WebAssembly and inject that using SEA. See https://nodejs.org/api/wasi.html

**Reply by mhdawson** [reply]
To add some clarification.

1) The existing functionality allows you to inject a JavaScript file
2) You can make the JavaScript in that file load/execute WebAssembly with WASI with the example above.

**Reply by krpeacock** [reply]
I may be reading this wrong, but isn't that the inverse of what I'm asking for? My interest would be to compile a Node.js application into a WebAssembly module that could be loaded by any wasm runtime.

**Reply by mhdawson** [reply]
Ah yes that is something quite different and I'm not sure how that relates to Single executable applications as we've been discussing them.

**Reply by RaisinTen** [reply]
I also don't think that this relates to the current state of SEAs, so I'll move this to the core repo discussions.

**Comment by mcollina** [comment]
@krpeacock I think adding a wasm/wasi build target for Node.js would be a major endeavor. We know it's kind of possible given that the private fork of Node.js in StackBlitz/WebContainer is doing just that, however, we'd need somebody to lead this effort.

**Comment by GabenGar** [comment]
> High performance: WebAssembly is designed to be fast and efficient, which makes it ideal for running code that requires high-performance, such as machine learning or scientific computing.

The major work of ML revolves around rattling the database/microservices constantly and compiling the functions which do this in WASM will not help it whatsoever.

The surprising part you didn't mention is wasm code cannot be codesplit/chunked because there is no such thing as dynamic import in wasm. And that means a potential wasm-ready SEA must be written in a very specific way incompatible with the way javascript programs are written. This can be a major issue for anyone working towards implementing it, since they will have to basically develop a javascript dialect, eat up all the problems caused by inventing another dialect and maintain it long-term.
